man/vh.1
.TH vh "1" "January 2025" verlihub "January 2025"
.SH NAME
vh \- This is a utility script for Verlihub
.SH OPTIONS
.TP
.BR "-i, --install [sleep_time]"
The option allows you to install a new hub. For detailed explanation please read <https://github.com/verlihub/verlihub/wiki/>
.TP
.BR "-uh, --uphelp [path_to_dbconfig]"
Update help files of hub at specified dbconfig path
.TP
.BR "-mp, --modport <path_to_dbconfig> <main> [extra]"
Modify hub ports
.TP
.BR "-r, --run [path_to_dbconfig]"
This option allows you to start the hub and run Verlihub as a daemon. Verlihub config folder can be omitted. If folder is omitted then the script will automatically try to guess them; the following folders will be searched in this order:
.RS
.TP
.BR " o value of VERLIHUB_CFG enviroment variable:"
.TP
.BR " o /etc/verlihub"
.TP
.BR " o PREFIX/etc/verlihub where PREFIX is the path where you installed Verlihub"
.TP
.BR " o ./.verlihub"
.TP
.BR " o ~/.verlihub (~ means your home directory)"
.P
If no directory will be found, the script will return an error.
.br
.I Example:
.nf
vh --run /etc/verlihub
.fi
.RE
.TP
.BR "-q, --restart [path_to_dbconfig]"
This option allows you to restart a running hub. You can also provide a Verlihub config folder; please refer to option --run for more information.
.TP
.BR "-s, --stop [path_to_dbconfig]"
This option allows you to stop a running hub. You can also provide a Verlihub config folder; please refer to option --run for more information
.TP
.BR "-t, --status [path_to_dbconfig]"
This option allows you to show if hub is running or not. You can also provide a Verlihub config folder; please refer to option --run for more information
.TP
.BR "-g, --geoip"
This option allows you to update all free MaxMind databases
.TP
.BR "-a, --autoload [path_to_dbconfig]"
This option allow you to create symbolic link in the hub config folder to load plugins automatically with Verlihub. You have to specify one the following options:
.RS
.TP
.BR "--list"
show a list of installed plugins and their status
.TP
.BR "--on"
load plugin automatically
.TP
.BR "--off"
disabled autoload feature of plugin
.P
You can specify the config directory using --configdir option; this is optional, please refer to option --run for more information.
.RE
.TP
.BR "-u, --adduser <nickname> <password> <class> [path_to_dbconfig]"
This option allows you to add a new user to the hub. You have to specify the following extra options to create a new user:
.RS
.TP
.BR "<nickname>"
the nickname of the user
.TP
.BR "<password>"
the password of the user. This is optional and it can be provided by the user.
.TP
.BR "<class>"
the class of the user. Allowed values are -1, 0, 1, 2, 3, 4, 5, 10. Please refer to manager the hub manual for more information about those classes.
.TP
.BR "[path_to_dbconfig]"
Verlihub config folder. This is optional; please refer to option --run for more information.
.RE
.TP
.BR "--clean [path_to_dbconfig]"
This option will ask you if you want to clean up kicklist, banlist, unbanlist, pi_iplog, pi_messages and pi_stats tables (plugins tables will be dropped only if they exists) using TRUNCATE method. This means that all data in these tables will be removed.
.TP
.BR "--backup <destination_path> [path_to_dbconfig]  "
This option allows you to make a backup of database that Verlihub uses to store registered users, triggers, configuration, etc. You have to specify a folder where you want to store the backup.
For 
.I example
if you want to save it to /home/foobar type:
.nf
vh --backup /home/foobar
.fi
You can also provide a Verlihub config folder; please refer to option --run for more information.
.TP
.BR "--restore <path_to_sql> [path_to_dbconfig] "
This option allows you to restore a backup created with option --backup or with other MySQL utility. You have to specify the path to sql file that contains the backup. 
For
.I example
if you want to restore verlihub_backup.sql that is located at /home/foobar type:
.nf
vh --restore /home/foobar/verlihub_backup.sql
.fi
You can also provide a Verlihub config folder; please refer to option --run for more information.
.TP
.BR "--version"
This option prints version information about this script and Verlihub.
.TP
.BR --help
displays help message with commands list
.SH AUTHOR
2003-2005 Daniel Muller, dan at verliba dot cz
2006-2025 Verlihub Team, info at verlihub dot net
.SH REPORTING BUGS
Official website: <https://github.com/verlihub/>
.br
Wiki pages: <https://github.com/verlihub/verlihub/wiki/>
.br
Support hub: <nmdcs://hub.verlihub.net:7777>
.br
Translate Verlihub using Transifex: <https://explore.transifex.com/feardc/verlihub/>
.SH COPYRIGHT
Verlihub is free software; You can redistribute it
and modify it under the terms of the GNU General
Public License as published by the Free Software
Foundation, either version 3 of the license, or at
your option any later version.
.br
Verlihub is distributed in the hope that it will be
useful, but without any warranty, without even the
implied warranty of merchantability or fitness for
a particular purpose. See the GNU General Public
License for more details.
.br
Please see https://www.gnu.org/licenses/ for a copy
of the GNU General Public License.


CHANGELOG
[ 1.6.1.0 ]

Fixed: Multithreading got locked by TLS proxy
Added: More useful output on CMake configuration

[ 1.6.0.0 ]

Added: Built-in TLS proxy instead of standalone command line utility, requires Go
Added: New vh command line utility options --getconf and --setconf

[ Older ]

Please follow commit log: https://github.com/verlihub/verlihub/commits/master


plugins/perl/vh/vh.pm
#	Copyright (C) 2003-2005 Daniel Muller, dan at verliba dot cz
#	Copyright (C) 2006-2025 Verlihub Team, info at verlihub dot net
#
#	Verlihub is free software; You can redistribute it
#	and modify it under the terms of the GNU General
#	Public License as published by the Free Software
#	Foundation, either version 3 of the license, or at
#	your option any later version.
#
#	Verlihub is distributed in the hope that it will be
#	useful, but without any warranty, without even the
#	implied warranty of merchantability or fitness for
#	a particular purpose. See the GNU General Public
#	License for more details.
#
#	Please see http://www.gnu.org/licenses/ for a copy
#	of the GNU General Public License.

package vh;

use 5.008;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration use vh ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	Ban
	CloseConnection
	GetMyINFO
	AddRegUser
	DelRegUser
	RegBot
	UnRegBot
	AddRobot
	DelRobot
	EditBot
	GetUserClass
	GetUserHost
	GetUserIP
	ParseCommand
	SendToClass
	SendToAll
	SendPMToAll
	KickUser
	SendDataToUser
	SetConfig
	GetConfig
	GetUsersCount
	GetNickList
	GetNickListArray
	GetOPList
	GetOPListArray
	GetBotList
	GetBotListArray
	GetBots
	GetTotalShareSize
	GetVHCfgDir

	SendToUser
	SendDataToAll
	Disconnect
	DisconnectByName

	VHDBConnect
	SQLQuery
	SQLFetch
	SQLFree

	GetUpTime
	IsBot
	IsUserOnline
	GetHubIp
	GetHubSecAlias
	StopHub
	GetTopic
	SetTopic
	GetIPCC
	GetIPCN
	InUserSupports
	ReportUser
	SendToOpChat
	ScriptCommand
	GetTempRights
	SetTempRights
) ] );

our @EXPORT_OK = ( "VH__Call__Function", @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
);

our $VERSION = '0.01';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&vh::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#XXX	if ($] >= 5.00561) {
#XXX	    *$AUTOLOAD = sub () { $val };
#XXX	}
#XXX	else {
	    *$AUTOLOAD = sub { $val };
#XXX	}
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('vh', $VERSION);

our $DBI_FOUND = 0;
eval 'use DBI; $DBI_FOUND=1;';

# Preloaded methods go here.

sub SendToUser {
  return SendDataToUser(@_);
}

sub SendDataToAll {
  return SendToClass(@_);
}

sub Disconnect {
  return CloseConnection(@_);
}

sub DisconnectByName {
  return CloseConnection(@_);
}

sub GetBots {
  carp "PerlScript plugin doesn't support GetBots!";
}

sub GetTempRights {
  carp "PerlScript plugin doesn't support GetTempRights currenlty";
}

sub SetTempRights {
  carp "PerlScript plugin doesn't support SetTempRights currenlty";
}

sub AddRobot {
  return RegBot(@_);
}

sub DelRobot {
  return UnRegBot(@_);
}

sub GetNickListArray {
  my $nl = GetNickList();
  $nl =~ s#^\$NickList\s+##;
  $nl =~ s#\$\$$##;
  my @nl = split /\$\$/, $nl;
  return @nl;
}

sub GetOPListArray {
  my $nl = GetOPList();
  $nl =~ s#^\$OpList\s+##;
  $nl =~ s#\$\$$##;
  my @nl = split /\$\$/, $nl;
  return @nl;
}

sub GetBotListArray {
  my $nl = GetBotList();
  $nl =~ s#^\$BotList\s+##;
  $nl =~ s#\$\$$##;
  my @nl = split /\$\$/, $nl;
  return @nl;
}

sub VHDBConnect {
  if (!$DBI_FOUND) {
    carp "No DBI found in your system. Install perl modules 'DBI' and 'DBD::mysql'";
    return;
  }

  use AppConfig;
  my $config = AppConfig->new({ERROR=>sub{}});

  for my $k(qw(db_host db_user db_pass db_data db_charset)) {
    $config->define("$k=s");
  }

  $config->file(vh::GetVHCfgDir()."/dbconfig");

  my $dsn = "DBI:mysql:database=".$config->db_data;
  if ($config->db_host) { $dsn.=";hostname=".$config->db_host; }
  my $dbh = DBI->connect($dsn, $config->db_user, $config->db_pass);
  $dbh->{'mysql_auto_reconnect'} = 1;
  if($config->db_charset) { $dbh->do("SET NAMES ".$config->db_charset); }

  return $dbh;
}

sub VH__Call__Function {
  my $func = shift;
  my $ref;
  my $ret;
  if (!defined &{"main::$func"}) {
    return 1;
  }
  no strict 'refs';
  return &{"main::$func"}(@_);
  use strict 'refs';
  return $ret;
}

BEGIN {
  require vh::const;
  vh::const->import();
}

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

vh - Perl extension for use in Verlihub Perl scripts.

=head1 SYNOPSIS

Sample script:

  use vh;

  my $dbh;
  my $botname;

  sub Main {
    $dbh = vh::VHDBConnect;

    $botname = vh::GetConfig("config", "hub_security");

    return 1;
  }

  sub VH_OnOperatorCommand {
    my $nick = shift;
    my $data = shift;
  
    if ($data =~ /^\!regscount\s*$/) {
      my $sth = $dbh->prepare("SELECT class, COUNT(nick) FROM reglist GROUP BY class ORDER BY class ASC");
      $sth->execute;

      my $s = "<$botname>\n\n";
      while(my ($c,$q) = $sth->fetchrow_array) {
        $s .= "class:$c users:$q\n";
      }
      $sth->finish;

      vh::SendToUser($s."|", $nick);
      return 0;
    }
    return 1;
  }

=head1 ABSTRACT

  Perl extension for use in Verlihub Perl scripts.

=head1 DESCRIPTION

You may use these callbacks from your scripts.

  //TODO&FIXME

  bool Ban(char *nick, char *op, char *nick, unsigned howlong, int bantype)
  bool CloseConnection(char *nick)
  const char *GetMyINFO(char *nick)
  char *ParseCommand(char *command_line)
  bool SendDataToAll(char *data, int min_class, int max_class)
  bool SendDataToUser(char *data, char *nick)

=head1 SEE ALSO

See PerlScript plugin documentation and demo scripts for more information.

=head1 AUTHORS

Daniel Muller E<lt>dan at verliba dot czE<gt>
Verlihub Team E<lt>info at verlihub dot netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2005 Daniel Muller, dan at verliba dot cz
Copyright (C) 2006-2025 Verlihub Team, info at verlihub dot net

Verlihub is free software; You can redistribute it
and modify it under the terms of the GNU General
Public License as published by the Free Software
Foundation, either version 3 of the license, or at
your option any later version.

Verlihub is distributed in the hope that it will be
useful, but without any warranty, without even the
implied warranty of merchantability or fitness for
a particular purpose. See the GNU General Public
License for more details.

Please see http://www.gnu.org/licenses/ for a copy
of the GNU General Public License.

=cut
